module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+	logger := e.log.With().Str("process", "notifications").Logger()
+	for {
+		next, blockID, ok := e.notificationState.NextNotification()
+		if !ok {
+			logger.Debug().Msg("no more notifications to process")
+			missing := e.notificationState.MissingHeights()
+			logger.Debug().Msgf("%#v", missing)
+
+			logger.Debug().Msgf("next should be %d", e.notificationState.lastNotified+1)
+			return
+		}
+
+		logger.Debug().Msgf("notifying for block %d", next)
+
+		executionData, ok := e.cache.Get(next)
+		if !ok {
+			logger.Debug().Msgf("execution data not in cache for block %d", next)
+			// get it from disk
+			var err error
+			executionData, err = e.eds.Get(ctx, blockID)
+			if err != nil {
+				// At this point the data has been downloaded and validated, so it should be available
+				// should this just delete the data and refetch?
+				// TODO: handle this better
+				ctx.Throw(fmt.Errorf("failed to get execution data for block: %w", err))
+			}
+		}
+
+		// send notifications
+		e.notifyConsumers(executionData)
+
+		// update notification manifest that we've sent notifications for this block height, and cleanup the cache
+		e.notificationState.Notified(next)
+		e.cache.Delete(next)
+	}
+}
+
+func (e *executionDataRequesterImpl) notifyConsumers(executionData *state_synchronization.ExecutionData) {
+	e.consumerMu.RLock()
+	defer e.consumerMu.RUnlock()
+	for _, fn := range e.consumers {
+		fn(executionData)
+	}
+}
+
+// Bootstrap Methods
+
+func (e *executionDataRequesterImpl) bootstrap(ctx irrecoverable.SignalerContext) error {
+	err := e.loadSyncState(ctx)
+	if err != nil {
+		return err
+	}
+
+	// TODO: if this is a fresh start, we'll need to provide the latest sealed block height as the
+	// target to catch up to before starting the finalization loop
+
+	return e.checkDatastore(ctx, e.notificationState.lastReceived)
+}
+
+func (e *executionDataRequesterImpl) loadSyncState(ctx irrecoverable.SignalerContext) error {
+
+	// TODO: load this from disk
+
+	err := e.notificationState.Load()
+	if err != nil {
+		e.log.Error().Err(err).Msg("failed to load notification state. using default")
+		e.notificationState = &status{}
+		// TODO: check if error is not found or something else
+	}
+
+	if e.notificationState.halted {
+		e.log.Error().Msg("HALTING REQUESTER: requester was halted on a previous run due to invalid data")
+		close(e.stop)
+		return ErrRequesterHalted
+	}
+
+	// defaults to the start block when booting with a fresh db
+	if e.notificationState.lastNotified == 0 {
+		e.notificationState.lastNotified = e.rootBlock.Header.Height
+		e.log.Debug().Msgf("setting lastNotified to root block height: %d", e.notificationState.lastNotified)
+	}
+
+	// at boot, we only have the start block and last block notifications were sent for,
+	// then we scan through the db to find the other metrics.
+	e.notificationState.lastReceived = e.notificationState.lastNotified
+	e.notificationState.lastSealed = e.notificationState.lastNotified
```

Comment:
I think last notified actually is the simplest since it's effectively a marker for that highest consecutive height fetched.. That way if a node crashes or reboots, it will pick up from there and requeue any that are missing.